src/diffusers/models/t5_attention.py

```diff
@@ -0,0 +1,421 @@
+            self.decoders.append(lyr)
+
+        self.decoder_norm = T5LayerNorm(t5config.d_model)
+
+        self.post_dropout = nn.Dropout(p=t5config.dropout_rate)
+        self.spec_out = nn.Linear(t5config.d_model, t5config.input_dims, bias=False)
+
+    def encoder_decoder_mask(self, query_input, key_input, pairwise_fn=torch.mul):
+        mask = pairwise_fn(query_input.unsqueeze(-1), key_input.unsqueeze(-2))
+        return mask.unsqueeze(-3)
+
+    def forward(self, encodings_and_masks, decoder_input_tokens, decoder_noise_time):
+        batch, _, _ = decoder_input_tokens.shape
+        assert decoder_noise_time.shape == (batch,)
+
+        # decoder_noise_time is in [0, 1), so rescale to expected timing range.
+        conditioning_emb = get_timestep_embedding(
+            decoder_noise_time * self.config.t5config.max_decoder_noise_time,
+            embedding_dim=self.config.t5config.d_model,
+            max_period=self.max_decoder_noise_time,
+        )
+
+        conditioning_emb = self.conditioning_emb(conditioning_emb)
+
+        assert conditioning_emb.shape == (batch, self.config.t5config.d_model * 4)
+
+        seq_length = decoder_input_tokens.shape[1]
+
+        # If we want to use relative positions for audio context, we can just offset
+        # this sequence by the length of encodings_and_masks.
+        decoder_positions = torch.broadcast_to(
+            torch.arange(seq_length, device=decoder_input_tokens.device),
+            (batch, seq_length),
+        )
+
+        position_encodings = self.position_encoding(decoder_positions)
+
+        # decoder: No padding present.
+        decoder_mask = torch.ones(decoder_input_tokens.shape[:2], device=decoder_input_tokens.device)
+
+        # Translate encoding masks to encoder-decoder masks.
+        encodings_and_encdec_masks = [(x, self.encoder_decoder_mask(decoder_mask, y)) for x, y in encodings_and_masks]
+
+        inputs = self.continuous_inputs_projection(decoder_input_tokens)
+
+        inputs += position_encodings
+
+        y = self.dropout(inputs)
+
+        # cross attend style: concat encodings
+        encoded = torch.cat([x[0] for x in encodings_and_encdec_masks], dim=1)
+        encoder_decoder_mask = torch.cat([x[1] for x in encodings_and_encdec_masks], dim=-1)
+        for lyr in self.decoders:
+            y = lyr(
+                y,
+                conditioning_emb=conditioning_emb,
+                encoder_hidden_states=encoded,
+                encoder_attention_mask=encoder_decoder_mask,
+            )[0]
+
+        y = self.decoder_norm(y)
+        y = self.post_dropout(y)
+
+        spec_out = self.spec_out(y)
+        return spec_out
+
+
+class ContinuousContextTransformer(ModelMixin, ConfigMixin):
+    @register_to_config
+    def __init__(self, t5config: T5Config):
+        super().__init__()
+
+        self.token_encoder = TokenEncoder(t5config=t5config)
+        self.continuous_encoder = ContinuousEncoder(t5config=t5config)
```

Comment:
I would actually seperate this into 3 model classes:

- 1) The `TokenEncoder` 
- 2) The `ContinuousEncoder` 
- 3) The `Decoder`

From my understanding only the `Decoder` is a diffusion model so I would say we move the classes `TokenEncoder` and `ContinuousEncoder` into `pipelien_spectrogram_diffusion.py` and then we code up the `Decoder` (maybe we can call it `T5Decoder`) in this file. 

Note that we cannot have a dependency on `transformers` in `src/diffusers/models` so I think we should code up the `Decoder` class from scratch here (copy paste `T5Block` from transformers. 

The `TokenEncoder` and `ContinuousEncoder` can heavily rely on `transformers` as it's totally fine to have a transformers dependency in the pipeline class.

Reply:
Also I'd propose to not make use of `T5Config` for the `Decoder` class but instead just write out all the kwargs (just as is done here: https://github.com/huggingface/diffusers/blob/c3dcb6749b744bbf3ebfd1329de6a50009b849ee/src/diffusers/models/attention.py#L87 ) so that we can correctly inherit from `ModelMixin` and `ConfigMixin` (otherwise as you mentioned you get a weird `self.config.config` problem. 

For `TokenEncoder` and `ContinuousEncoder` I'd inherit from `transformer`'s `PreTrainedModel and then you can pass the `config` just as done here.

Reply:
got it!